### PR TITLE
Implement UNet model option for segmentation

### LINKED ISSUE #114: Implement UNet model option for segmentation
Allow choosing which model is desired, mask rcnn or unet 

Reference for implementation: https://github.com/milesial/Pytorch-UNet

Additional reference: 
https://github.com/mberkay0/pretrained-backbones-unet

### PR DESCRIPTION
**Pull request by abuzarmahmood**
## Summary
This PR implements UNet as an alternative segmentation model alongside the existing Mask R-CNN implementation, addressing issue #114.

## Changes
- Added UNet model implementation based on [Pytorch-UNet](https://github.com/milesial/Pytorch-UNet)
- Created UNet-specific utilities for training and prediction
- Updated training script to support model selection via `--model` flag
- Updated prediction script to support model selection via `--model-type` flag
- Modified prediction_utils to handle both model architectures
- Added MODEL_SELECTION.md documentation comparing both models

### Training
```bash
# Train with Mask R-CNN (default)
python -m autoslide.src.pipeline.model.training

# Train with UNet
python -m autoslide.src.pipeline.model.training --model unet
```

### Prediction
```bash
# Predict with Mask R-CNN (default)
python -m autoslide.src.pipeline.model.prediction

# Predict with UNet
python -m autoslide.src.pipeline.model.prediction --model-type unet
```

## Model Comparison
- **Mask R-CNN**: Instance segmentation, better for individual vessel detection, more accurate but slower
- **UNet**: Semantic segmentation, faster inference, simpler architecture, better for dense segmentation

## Files Changed
- `autoslide/src/pipeline/model/unet_model.py` - UNet architecture implementation
- `autoslide/src/pipeline/model/unet_utils.py` - UNet training and prediction utilities
- `autoslide/src/pipeline/model/prediction_utils.py` - Updated to support both models
- `autoslide/src/pipeline/model/training.py` - Added model selection argument
- `autoslide/src/pipeline/model/training_utils.py` - Updated training loop for both models
- `autoslide/src/pipeline/model/prediction.py` - Added model type selection
- `MODEL_SELECTION.md` - Documentation for model selection

Resolves #114

### CODE REVIEW

**Comment by abuzarmahmood** [comment]
## Implementation Details

This implementation adds UNet as a model option while maintaining full backward compatibility with the existing Mask R-CNN implementation.

### Architecture Decisions

**UNet Model:**
- Standard U-Net architecture with encoder-decoder structure
- 5 downsampling and 4 upsampling blocks
- Skip connections between encoder and decoder
- Binary Cross-Entropy with Logits loss for training
- Adam optimizer (lr=0.001) for better convergence on this architecture

**Mask R-CNN Model:**
- Unchanged from existing implementation
- ResNet-50 backbone with FPN
- SGD optimizer (lr=0.005, momentum=0.9)

### Key Features

1. **Model Selection**: Users can choose between models via command-line arguments
2. **Separate Model Files**: Each model saves to its own file (`best_val_unet_model.pth` vs `best_val_mask_rcnn_model.pth`)
3. **Unified Interface**: Both models use the same prediction interface with automatic handling of architecture differences
4. **Data Loading**: UNet uses simpler data loading without augmentation for initial implementation

### Testing

All Python files compile successfully:
- ✅ `unet_model.py` - Syntax validated
- ✅ `unet_utils.py` - Syntax validated  
- ✅ `prediction_utils.py` - Syntax validated
- ✅ `training.py` - Syntax validated
- ✅ `training_utils.py` - Syntax validated
- ✅ `prediction.py` - Syntax validated

### Next Steps

Users can now:
1. Train both models and compare performance
2. Choose the model that best fits their use case (speed vs accuracy)
3. Switch between models for prediction without code changes

See `MODEL_SELECTION.md` for detailed usage instructions and model comparison.

**Comment by abuzarmahmood** [comment]
## Test Updates

Added comprehensive prediction tests that intelligently handle both model types.

### New Test File: `tests/test_prediction.py`

**Test Logic:**
- ✅ **Both models present**: Test both, fail if either fails
- ✅ **One model present**: Test that model, skip the other  
- ❌ **Model present but fails**: Fail the test
- ❌ **No models found**: Fail the test

### Test Classes

1. **TestModelInitialization** - Verify both architectures initialize correctly
2. **TestModelLoading** - Load models from disk with proper error handling
3. **TestPrediction** - Test prediction functionality for each model
4. **TestBothModels** - Comparative tests when both models available
5. **TestModelConsistency** - Verify deterministic output

### GitHub Workflow Updates

The `test-pipeline.yml` workflow now:
1. Checks which models are available before running predictions
2. Runs prediction with each available model independently
3. Fails if a present model fails (not just continues)
4. Skips if a model is not available
5. Fails if no models are found at all

### Example Output

**With only Mask R-CNN:**
```
✅ Mask R-CNN model found
❌ UNet model not found
✅ Running Mask R-CNN prediction
⚠️  Skipping UNet prediction (model not available)
```

**With both models:**
```
✅ Mask R-CNN model found
✅ UNet model found
✅ Running Mask R-CNN prediction
✅ Running UNet prediction
✅ Comparing model outputs
```

### Helper Scripts

- `verify_test_logic.py` - Check which models are available and preview test behavior
- `tests/README_PREDICTION_TESTS.md` - Detailed documentation of test behavior

This ensures the CI/CD pipeline properly validates both model implementations while gracefully handling cases where only one model is trained.
